// chathura-model/losses/MaeMsaHybrid.js
export default class HybirdMaeMse{
    validationLoss = undefined;
    isUseMsa = true;
    constructor({delta = false , useDefalut = false}={}){
        this.delta = delta;
        this.useDefalut = useDefalut && !this.delta;
    }

    forword(y_pred , y_true){
        this.y_pred = y_pred;
        this.y_true = y_true;
        this.loss = Math.abs(y_pred - y_true)

    }
    backword(dvalues = 1){
        this.input_dvalues = dvalues;

        if(this.validationLoss && this.delta && this.delta > this.validationLoss) this.isUseMsa = false;
        // if(this.useDefalut && Number.isFinite(this.loss ** 2)) isUseMsa = false  

        // console.log(isUseMsa,this.loss)
    if(this.isUseMsa){
        if(this.y_pred[0][0] > this.y_true[0]){
            this.dloss = 1;
        }else if(this.y_pred[0][0] < this.y_true[0] ){
            this.dloss = -1
        }
        this.dloss = [[this.dloss * this.input_dvalues]]
    }else{
        this.dloss = [[2*(this.y_pred - this.y_true)]];
    }
    }
}
